### PR TITLE
Processing Refactoring and more

### PR DESCRIPTION
**Pull request by profxj**
This PR began with the goals to enable: (1) generation of inverse variance arrays,
masks, etc. for calibration images; (2) use counts instead of ADU in calibration images;
(3) use the same algorithm as the science image to combine.
It accomplishes that and a bit more.  

Here are the main CHANGES:

- Added pypeit_chk_flat script to view flat images
- Refactored image objects into `RawImage`, `ProcessRawImage`, `PypeItImage`, `BuildImage`
- Moved load() and save() methods from `MasterFrame` to the individual calibration objects
- Converted ArcImage and FlatImage images into counts
- Added code to allow for IVAR and RN2 image generation for calibs
- Added several from_master_file() instantiation methods
- Use coadd2d.weighted_combine() to stack calibration images

Note that several of the image objects are intentionally light-weight (i.e. they
are simple containers for 3+ images) and several are instantiated and then
quickly destroyed (`ProcessRawImage`, `BuildImage`).

The changes to `MasterFrame` means some duplication of code in the 
calibration objects but I consider this a fair price to pay compared with
the increasing level of complexity in the load() and save() methods to handle
more complex calibration objects.

Documentation has been polished throughout.

### CODE REVIEW

**Comment by profxj** [comment]
--- PYPEIT DEVELOPMENT SUITE FAILED 1/30 TESTS ---

The one test fail is in slit definition for keck_lris_blue_multi_300_5000_d680.

I have confirmed that test passes in the `trace_testing` branch 
which will follow this PR and overhauls slit edge analysis.
I am therefore not inclined to fix it at this time.

File: pypeit/biasframe.py [hunk 1]
```diff
@@ -50,6 +52,33 @@ class BiasFrame(calibrationimage.CalibrationImage, masterframe.MasterFrame):
...
+            master_file (str):
+            par (:class:`pypeit.par.pypeitpar.FrameGroupPar`, optional):
+
+        Returns:
+            biasframe.BiasFrame:
+                The PypeItImage is loaded into self.pypeitImage
+
+        """
+        # Spectrograph
+        spectrograph, extras = masterframe.items_from_master_file(master_file)
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/biasframe.py]: Let me add that when I'm developing, I find the number of lines of code required to read in a master file super annoying. For that reason, I think having utility functions that process the master file output into the required products in one line of code is very valuable. Whether these should be class methods, or just methods in the same module is not totally clear to me, and probably does not matter.

File: pypeit/images/buildimage.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+        # Process
+        processrawImage = processrawimage.ProcessRawImage(rawImage, self.par, bpm=bpm)
+        processedImage = processrawImage.process(process_steps, bias=bias, pixel_flat=pixel_flat,
+                                                 illum_flat=illum_flat)
+
+        # Return
+        return processedImage
+
+    def run(self, process_steps, bias, pixel_flat=None, illum_flat=None,
```
Comment (jhennawi) [inline pypeit/images/buildimage.py]: Why is pixel_flat a required argument above, but not a required argument here? In my opinion, only filename should be required. Process steps and bias should be optional. Process_steps should default to just trimming, Oscan_subtraction and gain multiplication. 

I've been begging for the equivalent of 

IDL> long_proc, filename, imag, ivar 

for quite some time now, but never got my wish. Can you make that a reality please in this PR. We can debate what should be optional vs or required, but I think there really needs to be a one line utility function for this class.

File: pypeit/images/calibrationimage.py [hunk 3]
```diff
@@ -132,42 +119,19 @@ def build_image(self, bias=None, bpm=None):
...
-            self.image = combine.comb_frames(image_arr,
-                                         saturation=self.spectrograph.detector[self.det-1]['saturation'],
-                                         method=self.proc_par['combine'],
-                                         satpix=self.proc_par['satpix'],
-                                         cosmics=self.proc_par['sigrej'],
-                                         n_lohi=self.proc_par['n_lohi'],
-                                         sig_lohi=self.proc_par['sig_lohi'],
-                                         replace=self.proc_par['replace'])
+        buildImage = buildimage.BuildImage(self.spectrograph, self.det, self.proc_par, self.file_list)
+        self.pypeitImage = buildImage.run(self.process_steps, bias, bpm=bpm, ignore_saturation=ignore_saturation)
```
Comment (jhennawi) [inline pypeit/images/calibrationimage.py]: Thanks for cleaning this up!

File: pypeit/wavetilts.py [hunk 4]
```diff
@@ -417,71 +437,35 @@ def load(self, ifile=None, return_header=False):
...
-        if not os.path.isfile(_ifile):
-            # Master file doesn't exist
-            msgs.warn('No Master {0} frame found: {1}'.format(self.master_type, self.file_path))
-            return empty_return
-
-        # Read and return
-        msgs.info('Loading Master {0} frame: {1}'.format(self.master_type, _ifile))
-        return self.load_from_file(_ifile, return_header=return_header)
-
-    @staticmethod
```
Comment (jhennawi) [inline pypeit/wavetilts.py]: See my comment above. I don't want to load in the WaveTilts class from file as you have now changed things. I just want the tilts_dict. The load_from_file method below is a simple one line interface to read in the master from the file. The WaveTilts object is not used in PypeIt as a master, but the tilts_dict is. I don't need to instantiate a class, I want to read in that dict. Please add load_from_file functionality back to this class.

File: pypeit/arcimage.py [hunk 5]
```diff
@@ -82,12 +87,18 @@ def save(self, outfile=None, overwrite=True):
...
-    def load(self, ifile=None, return_header=False):
+        _outfile = self.master_file_path if outfile is None else outfile
+        # Check if it exists
+        if os.path.exists(_outfile) and not overwrite:
+            msgs.warn('Master file exists: {0}'.format(_outfile) + msgs.newline()
+                      + 'Set overwrite=True to overwrite it.')
+            return
+        #
+        hdr = self.build_master_header(steps=self.process_steps, raw_files=self.file_list)
+        pypeitimage.save_images(self.pypeitImage, _outfile, hdr=hdr, iext='ARC')
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/arcimage.py]: Only because I had/have this philosophy to keep this object super light-weight.
Almost a container and nothing further.

I'm won't be adamant about it.  What are the pros/cons here?

[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/arcimage.py]: Having stewed on it a bit further and reminding myself that the `ImageMask` code is already inherited, I am going to push these into the class too.  Will still try to keep it quite light-weight but we'll see if we can toe that line.

File: pypeit/arcimage.py [hunk 6]
```diff
@@ -102,5 +113,11 @@ def load(self, ifile=None, return_header=False):
             Returns a `numpy.ndarray`_ with the arc master frame image.
             Also returns the primary header, if requested.
         """
-        return super(ArcImage, self).load('ARC', ifile=ifile, return_header=return_header)
+        # Check on whether to reuse and whether the file exists
+        master_file = self.chk_load_master(ifile)
+        if master_file is None:
+            return
+        else:  # Load
+            self.pypeitImage = pypeitimage.load_images(master_file)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/arcimage.py]: Ok, done.  Renamed `from_file()`

File: pypeit/biasframe.py [hunk 7]
```diff
@@ -108,56 +139,58 @@ def save(self, outfile=None, overwrite=True):
             overwrite (:obj:`bool`, optional):
                 Overwrite any existing file.
         """
-        if self.image is None:
+        # Some checks
+        if self.pypeitImage is None:
             msgs.warn('No MasterBias to save!')
             return
-        if not isinstance(self.image, np.ndarray):
+        if not isinstance(self.pypeitImage.image, np.ndarray):
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/biasframe.py]: Agreed.  Worth having even as I try to maintain this being a very lightweight object.

File: pypeit/arcimage.py [hunk 8]
```diff
@@ -102,5 +113,11 @@ def load(self, ifile=None, return_header=False):
             Returns a `numpy.ndarray`_ with the arc master frame image.
             Also returns the primary header, if requested.
         """
-        return super(ArcImage, self).load('ARC', ifile=ifile, return_header=return_header)
+        # Check on whether to reuse and whether the file exists
+        master_file = self.chk_load_master(ifile)
+        if master_file is None:
+            return
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/arcimage.py]: Unless one has been previously loaded, it will be `None`.
And I thought it best to not over-write what was in the attribute
on a failed load.

File: pypeit/core/load.py [hunk 9]
```diff
@@ -564,3 +564,35 @@ def waveids(fname):
         pass
     return pixels
 
+
+def load_multiext_fits(filename, ext):
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/load.py]: Definitely not necessary, but I find the code cleaner where it is used
than re-creating this method in those 3 places.  With more to come..

File: pypeit/core/tracewave.py [hunk 10]
```diff
@@ -369,6 +369,8 @@ def trace_tilts_work(arcimg, lines_spec, lines_spat, thismask, slit_cen, inmask=
     nuse = np.sum(use_tilt)
 
     msgs.info('Number of usable arc lines for tilts: {:d}/{:d}'.format(nuse,nlines))
+    if nuse == 0:
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/tracewave.py]: removed

File: pypeit/flatfield.py [hunk 11]
```diff
@@ -300,29 +337,44 @@ def save(self, outfile=None, overwrite=True):
...
-            illumination flat.  Also returns the primary header, if
-            requested.
+            illumination flat.
         """
-        return super(FlatField, self).load(['RAWFLAT', 'PIXELFLAT', 'ILLUMFLAT'], ifile=ifile,
-                                           return_header=return_header)
+        # Check on whether to reuse and whether the file exists
+        master_file = self.chk_load_master(ifile)
+        if master_file is None:
+            return (None, None, None)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/flatfield.py]: gone

File: pypeit/images/buildimage.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+from IPython import embed
+
+
+class BuildImage(object):
+    """
+    Class to generate an image from one or more files (and other pieces).
+
+    The core processing steps are handled by ProcessRawImage
+    This object is mainly for combining multiple images
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/buildimage.py]: Going with `CombineImage`

File: pypeit/images/buildimage.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    @property
+    def nfiles(self):
+        """
+        Number of files in the files attribute
+
+        Returns:
+            int
+
+        """
+        if isinstance(self.files, list):
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/buildimage.py]: so clean

File: pypeit/images/calibrationimage.py [hunk 14]
```diff
@@ -7,71 +7,58 @@
...
-from pypeit.images import maskimage
-
-
+from pypeit.images import buildimage
 
 from IPython import embed
 
 
-class CalibrationImage(pypeitimage.PypeItImage, maskimage.ImageMask):
+class CalibrationImage(object):
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/calibrationimage.py]: Because some of the `CalibrationImage` classes (e.g. `ArcImage`) have evolved to becoming the objects which generate a `PypeItImage`, while others generate entirely different things or mutliple `PypeItImage` objects (e.g. `FlatField`).  

I nearly eliminated this `CalibrationImage` object but decided that I didn't want to duplicate the `build_image()` method in the several places it is used.  And the file checking methods are somewhat useful.

File: pypeit/images/pypeitimage.py [hunk 15]
```diff
@@ -1,49 +1,65 @@
...
+
+from pypeit.core import save
+from pypeit.images import maskimage
+from pypeit.io import initialize_header
+
 from IPython import embed
 
-class PypeItImage(object):
+
+class PypeItImage(maskimage.ImageMask):
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/pypeitimage.py]: If it becomes an attribute, we'll be accessing it with syntax like  `pypeitImage.mask.mask` and `pypeitImage.mask.bpm`.  Not a disaster, but I don't see a downside of inheriting.

As a side note, I find myself deciding that given the `PypeItImage` object already inherits the mask code then we might as well tag on the save/load methods.  And so it begins to move from being light-weight..   :)

File: pypeit/images/buildimage.py (hunk 2 above)
Comment (profxj) [reply pypeit/images/buildimage.py]: None of the calibration images use a pixel_flat.  Have made it optional above now.

As for your IDL request, I have a hard time parsing such code, but I believe this does the trick:

```
In [1]: from pypeit.spectrographs import util                                  
In [2]: shaneb = util.load_spectrograph('shane_kast_blue')                     
In [3]: from pypeit.images.processrawimage import process_raw_for_jfh          
In [4]: pypeitImage = process_raw_for_jfh('b13.fits.gz', shaneb)               
[WARNING] :: processrawimage.py 397 process_raw_for_jfh() - Using the Processinng parameters from scienceframe
```

I look forward to hearing how much easier this makes your coding.

File: pypeit/images/calibrationimage.py (hunk 3 above)
Comment (profxj) [reply pypeit/images/calibrationimage.py]: Felt good

File: pypeit/images/pypeitimage.py [hunk 16]
```diff
@@ -1,49 +1,65 @@
...
         head0 (astropy.io.fits.Header):
-        orig_shape (tuple):
-        binning_raw (tuple):  Binning in the raw image orientaion (NAXIS1, NAXIS2)
-        binning (tuple): Binning the PypeIt orientation (spec, spat)
-        exptime (float): Exposure time of the image
 
     """
 
-    def __init__(self, spectrograph, det):
+    def __init__(self, image, ivar=None, rn2img=None, bpm=None,
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/pypeitimage.py]: A `PypeItImage` is agnostic about the `spectrograph`
and I feel should also not care about the `det`

File: pypeit/images/scienceimage.py [hunk 17]
```diff
@@ -9,281 +11,59 @@
...
-            Read noise**2 image
-        datasec_img (np.ndarray):
-            Mapping of pixels to amplifier
-        filename (str):
-            Required to build from a Raw image
     """
     frametype = 'science'
 
-    def __init__(self, spectrograph, det, par, bpm):
+    def __init__(self, spectrograph, det, par, image, ivar, bpm, rn2img=None,
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/images/scienceimage.py]: Look at the code.  They are required.

File: pypeit/pypeit.py [hunk 18]
```diff
@@ -542,22 +542,22 @@ def extract_one(self, frames, det, bg_frames=[], std_outfile=None):
...
 
         # Background subtract?
         if len(bg_frames) > 0:
             bg_file_list = self.fitstbl.frame_paths(bg_frames)
-            self.sciImg = self.sciImg - scienceimage.ScienceImage.from_file_list(
+            self.sciImg = self.sciImg - scienceimage.build_from_file_list(
                 self.spectrograph, det, self.par['scienceframe']['process'],
                 self.caliBrate.msbpm, bg_file_list, self.caliBrate.msbias,
-                self.caliBrate.mspixelflat.copy(), illum_flat=self.caliBrate.msillumflat)
+                self.caliBrate.mspixelflat, illum_flat=self.caliBrate.msillumflat)
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/pypeit.py]: Because it got modified within the method.
It is now copied in the method.

**Comment by profxj** [comment]
Here is an Inheritance diagram for @kbwestfall 

I can't quite get the Sphinx to work, although I am close.

imageobjects

**Comment by profxj** [comment]
@jhennawi raises two issues related to loading a `MasterFrame`:

1.  He wants the key item for the object in a one-line call.  This is already
implemented.   See `coadd2d.py`

2.  He doesn't think the object needs to be instantiable from the master frame
output file.   @kbwestfall and I disagree.   And I'll note that several of the current
`from_master_file()` methods instantiate the object but without it having full 
functionality.  The bottom line on this, is that we will have one method to load
from a master file and it will instantiate the object.  And load the key output items
which one then uses in calibrations.   See the way it is done for `WaveImage`.

File: pypeit/waveimage.py [hunk 19]
```diff
@@ -168,11 +198,18 @@ def load(self, ifile=None, return_header=False):
...
+        #return super(WaveImage, self).load('WAVE', ifile=ifile, return_header=return_header)
+        master_file = self.chk_load_master(ifile)
+        if master_file is None:
+            return None
+        # Load
+        self.image, head0 = load.load_multiext_fits(master_file, ['WAVE'])
+        # Return
+        return self.image
 
+    '''
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/waveimage.py]: see discussion below

File: pypeit/biasframe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/biasframe.py]: It looks like we have 3 overlapping and differing opinions on this one.
I agreed the extras list in `items_from_master_file` could get unruly but 
as already have more than one way of saving master info (FITS, JSON)
I find it somewhat unavoidable.    If we eventually go to only FITS (possible)
this could be stripped down.

As for not opening the file more than once, I see that as good practice
but I don't feel it is worth the extra coding (e.g. it would include adding a new
method on `PypeItImage`.  

Bottom line -- I didn't change much on this one other than adding to the doc string.

File: pypeit/scripts/chk_flats.py [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    # Load up
+    flatField = flatfield.FlatField.from_master_file(pargs.master_file)
+
+    try:
+        ginga.connect_to_ginga(raise_err=True)
+    except ValueError:
+        subprocess.Popen(['ginga', '--modules=RC'])
+        time.sleep(3)
+
+    # Show RawFlatImage
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/scripts/chk_flats.py]: `flat_model` is currently not written to disk. Do we wish to add that?

And I can't find any QA in `flat.py` nor `flatfield.py`

File: pypeit/wavetilts.py [hunk 21]
```diff
@@ -62,12 +63,41 @@ class WaveTilts(masterframe.MasterFrame):
...
+        gpm (np.ndarray):
+            Good pixel mask
+            Eventually, we might attach this to self.msarc although that would then
+            require that we write it to disk with self.msarc.image
     """
-    
     # Frametype is a class attribute
     master_type = 'Tilts'
 
+    @classmethod
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/wavetilts.py]: See below

File: pypeit/wavetilts.py (hunk 4 above)
Comment (profxj) [reply pypeit/wavetilts.py]: See below

File: pypeit/scripts/chk_flats.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/scripts/chk_flats.py]: Take a  look at the show method in flatfield.py, line 267.  Yes, I think we should probably write the model to disk.

**Comment by jhennawi** [comment]
I would like to discuss point #2 above in our next telecon, and will wait until we have that discussion to a approve.

File: pypeit/images/buildimage.py (hunk 2 above)
Comment (jhennawi) [reply pypeit/images/buildimage.py]: Thanks for adding this. Even better would be a "header2spectrograph" method, that does not require the spectrograph.

File: pypeit/scripts/chk_flats.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/scripts/chk_flats.py]: Done

File: pypeit/images/buildimage.py (hunk 2 above)
Comment (profxj) [reply pypeit/images/buildimage.py]: I'll let you code that up for the 10+ spectrographs
and their unique headers.  Probably won't be too hard.

**Comment by jhennawi** [comment]
Hi @profxj, 

I'm happy with your PR. However, keck_lris_blue_multi_300_5000_d680 is failing the tests. You claim it passes in a new branch. However, this test was not failing before in develop. If it was, I can approve your PR, but I'm pretty sure this was not the case. So apparently something related to your changes to image processing is causing it to fail. That sounds to me like a bug. Please investigate.

**Comment by profxj** [comment]
It is related to slit tracing on the new slit image produced
by our combining, etc.

I am not going to fix this as the edge tracing code is fully revamped
in the next PR.  If it still fails there, we'll fix it.
